src/jemalloc.c

```diff
@@ -4148,8 +4170,41 @@ batch_alloc(void **ptrs, size_t num, size_t size, int flags) {
 					bin = &tcache->bins[ind];
 				}
 			}
+			/*
+			 * If we don't have a tcache bin, we don't want to
+			 * immediately give up, because there's the possibility
+			 * that the user explicitly requested to bypass the
+			 * tcache, and in such cases, we go through the slow
+			 * path, i.e. the mallocx() call at the end of the
+			 * while loop.  A slight overhead here would be that
+			 * the "if (bin == NULL)" branch above will be entered
+			 * in every iteration of the while loop, but it
+			 * shouldn't be a huge issue, because (a) the
+			 * computations there are cheap and (b) users
+			 * requesting to bypass the tcache care less about
+			 * performance.
+			 */
 			if (bin != NULL) {
 				size_t bin_batch = batch - progress;
+				/*
+				 * n can be less than bin_batch, meaning that
+				 * the cache bin does not have enough memory.
+				 * In such cases, we rely on the slow path,
+				 * i.e. the mallocx() call at the end of the
+				 * while loop, to fill in the cache, and in the
+				 * next iteration of the while loop, the tcache
+				 * will contain a lot of memory, and we can
+				 * harvest them here.  Compared to the
+				 * alternative approach where we directly go to
+				 * the arena bins here, the overhead of our
+				 * current approach should usually be minimal,
+				 * since the tcache is usually larger than a
```

Comment:
Is this true? I think lg_fill_div is often positive, so we fill less than a slab amount into tcaches.

Reply:
You're right.